### PR TITLE
make `noop_method_call` warn by default

### PR DESCRIPTION
**Pull request by fee1-dead**
r? @compiler-errors

### CODE REVIEW

File: tests/ui/lint/noop-method-call.stderr [hunk 1]
```diff
@@ -1,62 +1,58 @@
 warning: call to `.clone()` on a reference in this situation does nothing
-  --> $DIR/noop-method-call.rs:16:71
+  --> $DIR/noop-method-call.rs:15:71
    |
 LL |     let non_clone_type_ref_clone: &PlainType<u32> = non_clone_type_ref.clone();
    |                                                                       ^^^^^^^^ unnecessary method call
    |
    = note: the type `&PlainType<u32>` which `clone` is being called on is the same as the type returned from `clone`, so the method call does not do anything and can be removed
```
Comment (erikdesjardins) [inline tests/ui/lint/noop-method-call.stderr:7]: This note doesn't really make sense. `Vec<i32>::clone` also returns the same type it's called on, but that doesn't mean it does nothing.

I would suggest something like:
```suggestion
   = note: the type `PlainType<u32>` does not implement `Clone`, so calling `clone` on `&PlainType<u32>` clones the reference, which does not do anything and can be removed
```

File: compiler/rustc_lint/src/noop_method_call.rs [hunk 2]
```diff
@@ -76,21 +75,16 @@ impl<'tcx> LateLintPass<'tcx> for NoopMethodCall {
...
-                }
-                _ => return,
-            },
-            _ => return,
+        //
+        // Verify we are dealing with a method/associated function, and check that we're dealing
+        // with a trait method for one of the traits we care about.
+        let (did, trait_) = if let Some((DefKind::AssocFn, did)) = cx.typeck_results().type_dependent_def(expr.hir_id)
+        && let Some(trait_id) = cx.tcx.trait_of_item(did)
+        && let Some(trait_ @ (sym::Borrow | sym::Clone | sym::Deref)) = cx.tcx.get_diagnostic_name(trait_id)
```
Comment (compiler-errors) [inline compiler/rustc_lint/src/noop_method_call.rs:83]:
```suggestion
            && let Some(trait_id) = cx.tcx.trait_of_item(did)
            && let Some(trait_ @ (sym::Borrow | sym::Clone | sym::Deref)) = cx.tcx.get_diagnostic_name(trait_id)
```

**Review by compiler-errors — APPROVED** [review]
Does bumping a lint from allow -> warn need an fcp or anything? Implementation looks fine now -- I like the new rewording too.

**Comment by fee1-dead** [comment]
Not sure about whether a fcp is needed.

cc @rust-lang/compiler-contributors

**Comment by RalfJung** [comment]
> `cc @rust-lang/compiler-contributors`

(That's a huge ping group with more than 40 people, most of whom do not have FCP power.  It's probably not what you wanted to ping here.)

**Comment by fee1-dead** [comment]
It's the group that gets pinged for MCPs so I pinged.

**Comment by oli-obk** [comment]
This lint was made [allow-by-default in the original PR](https://github.com/rust-lang/rust/pull/80723#issuecomment-780135058), but there was never the proposed follow up to make it warn-by-default. I think it was mostly allow-by-default to make it easier for clippy to transition?

This implemented the lint proposed in the following lang-team MCP: https://github.com/rust-lang/lang-team/issues/67

The lang team would like to see a crater run with this lint denied to judge its effects on existing code.

**Comment by fee1-dead** [comment]
experiment PR now up at #112160

**Comment by fee1-dead** [comment]
Marking as S-waiting-on-team

**Comment by oli-obk** [comment]
> experiment [completed](https://github.com/rust-lang/rust/pull/112160#issuecomment-1578134257). Should I add a suggestion to the diagnostic?

Please summarize the crater results. You can then add the lang team nominated label so their triage picks it up

**Comment by joshtriplett** [comment]
As @oli-obk mentioned, we need a summary of the crater results; please mark this as S-waiting-on-team again when that's available.

(The most important data in such a summary would be whether any of the errors look like false positives.)

**Comment by fee1-dead** [comment]
I checked through all the results in the "regressed: dependencies" tab, while most were due to "No space left on device", these regressed due to the lint: [1](https://crater-reports.s3.amazonaws.com/pr-112160/try%23024378d9203a4009ed73d6db2681b88a8d74ce7a/gh/TheLocust3.simple-virtual-machine/log.txt) [2](https://crater-reports.s3.amazonaws.com/pr-112160/try%23024378d9203a4009ed73d6db2681b88a8d74ce7a/gh/bencecile.CompletionTracker/log.txt) [3](https://crater-reports.s3.amazonaws.com/pr-112160/try%23024378d9203a4009ed73d6db2681b88a8d74ce7a/gh/Akamitori.crypto_pals_challenges/log.txt) [4](https://crater-reports.s3.amazonaws.com/pr-112160/try%23024378d9203a4009ed73d6db2681b88a8d74ce7a/gh/jDomantas.fatfs-test/log.txt) [5](https://crater-reports.s3.amazonaws.com/pr-112160/try%23024378d9203a4009ed73d6db2681b88a8d74ce7a/gh/alexiwalker.mackley/log.txt) [6](https://crater-reports.s3.amazonaws.com/pr-112160/try%23024378d9203a4009ed73d6db2681b88a8d74ce7a/gh/kantic.termc/log.txt) [7](https://crater-reports.s3.amazonaws.com/pr-112160/try%23024378d9203a4009ed73d6db2681b88a8d74ce7a/gh/v4kst1z.TIP/log.txt) [8](https://crater-reports.s3.amazonaws.com/pr-112160/try%23024378d9203a4009ed73d6db2681b88a8d74ce7a/gh/v4kst1z.TIP/log.txt) [9](https://crater-reports.s3.amazonaws.com/pr-112160/try%23024378d9203a4009ed73d6db2681b88a8d74ce7a/gh/hsakas.rust-cli-todo/log.txt) [10](https://crater-reports.s3.amazonaws.com/pr-112160/try%23024378d9203a4009ed73d6db2681b88a8d74ce7a/gh/playXE.Jazz/log.txt). This one failed due to a cryptic linker error, so probably spurious: [11](https://crater-reports.s3.amazonaws.com/pr-112160/try%23024378d9203a4009ed73d6db2681b88a8d74ce7a/reg/marine-rs-sdk-test-0.10.0/log.txt).

I did a spot check of about 50 from the root results, and most of them look fine (not false positives). I did find that more than half of the crates that triggered the lint already had unused warnings and deprecated warnings, etc, where code quality is already not very high.

I notice that the warning on `s.clone()` when `type_of(s) == &str` could be improved, since it currently says that because `str` doesn't implement `Clone` only the reference is cloned, but we could probably say something that is clearer, since DSTs can't implement clone.

**Comment by fee1-dead** [comment]
Nominating for lang team discussion. Note that the crater experiment made the lint deny by default, while this PR is proposing to make it warn by default.

**Comment by estebank** [comment]
> I notice that the warning on s.clone() when type_of(s) == &str could be improved, since it currently says that because str doesn't implement Clone only the reference is cloned, but we could probably say something that is clearer, since DSTs can't implement clone. ([example lint message](https://crater-reports.s3.amazonaws.com/pr-112160/try%23024378d9203a4009ed73d6db2681b88a8d74ce7a/gh/JoshuaRadin37.lrmalloc.rs-testing/log.txt))

I believe that we can adjust the implementation to handle these cases more gracefully (potentially by not emitting the lint at all), either in general or at least for `&str` in particular.

**Comment by scottmcm** [comment]
@rfcbot fcp merge

From the random things I checked in the crater run, this looks like it'll be a great warning to have on-by-default.

**Comment by nikomatsakis** [comment]
@rfcbot reviewed

I've seen plenty of user confusion and I *think* seen bugs related to this.

**Comment by compiler-errors** [comment]
I don't think that this diagnostic has a suggestion, but I agree that it probably could.

**Comment by fee1-dead** [comment]
I have added a suggestion.

**Pull request merged**